File: LOL_IMG.py
import re
import requests
import os
import json
import threading
from multiprocessing import Pool


class LolPic:
    def __init__(self):
        self.hero_data = hero_data
        self._base_url = 'http://ossweb-img.qq.com/images/lol/web201310/skin/big'

    def get_hero_image_ids(self, hero_name):
        """
        从英雄对应的js文件中获取皮肤id和名称
        :param hero_name: 英雄的名称
        :return: skin_id_dict :一个字典列表，元素为皮肤对应的数据
        :retype: [{},{},...]
        """
        hero_js_url = 'https://lol.qq.com/biz/hero/{}.js'.format(hero_name)
        html = requests.get(hero_js_url)
        if html.status_code == 200:
            html.encoding = 'utf-8'
            skin_id_dict_str = re.findall('"skins":(.*?),"info"', html.text, re.S)[0]
            skin_id_dict_list = json.loads(skin_id_dict_str)
            yield from skin_id_dict_list  # type: [{}, {}]
        else:
            print(f'{hero_name}.js 请求出错')

    def download_hero_image(self, hero_name):
        """
        图片下载函数，通过调用self.get_hero_image_ids来获取英雄对应的数据，
        然后通过self.save_image来请求并保存图片
        :param hero_name :英雄的名字->Aatrox...
        :type :str
        :return None
        """
        print(f'\n{"*"*20}开始下载 {hero_name}{"*"*20}')
        threads = list()
        for skin_data in self.get_hero_image_ids(hero_name):
            hero_skin_url = self._base_url + skin_data["id"] + '.jpg'
            hero_skin_name = skin_data["name"]
            hero_skin_name = re.sub('[\\\\/:*?\"<>|]', '', hero_skin_name)
            threads.append(threading.Thread(target=self.save_img,
                                            args=(hero_skin_url, hero_skin_name, hero_name)))
        for thread in threads:
            thread.start()
        for thread in threads:
            thread.join()
        print(f'{"*"*20}{hero_name} 下载完成{"*"*20}\n')

    def save_img(self, hero_skin_url, hero_skin_name, hero_name):
        """
        图片保存函数
        :param hero_skin_url: 英雄皮肤链接
        :param hero_skin_name: 英雄皮肤名称
        :param hero_name: 英雄名字，用来生成对应的文件夹
        :return: None
        """
        response = requests.get(url=hero_skin_url)
        if os.path.exists(f'./lolimg/{hero_name}'):
            pass
        else:
            os.makedirs(f'./lolimg/{hero_name}')
        if response.status_code == 200:
            with open(f'./lolimg/{hero_name}/{hero_skin_name}.jpg', 'wb') as f:
                f.write(response.content)
            print(f'{hero_name} {hero_skin_name} 下载完成')
        else:
            print(f'{hero_name} {hero_skin_name} 下载失败，url为{hero_skin_url}')

    def main(self):
        """
        主函数，用来进行多进程调用，4个进程
        :return: None
        """
        pool = Pool(processes=4)
        for _, hero_name in self.hero_data.items():
            pool.apply_async(func=self.download_hero_image, args=(hero_name,))
        pool.close()
        pool.join()


if __name__ == '__main__':
    fp = open('./lol_hero.json', 'r', encoding='utf-8')
    hero_data = json.load(fp)  # 读取英雄的id和名称
    fp.close()
    hero_img = LolPic()
    hero_img.main()

File: lol_img.go
/* author: lyn
date: 2018/12/7
name: LOL_Img
*/

package main

import (
	"fmt"
	"github.com/json-iterator/go"
	"io/ioutil"
	"net/http"
	"os"
	"regexp"
	"runtime"
	"sync"
)

const baseUrl = "http://ossweb-img.qq.com/images/lol/web201310/skin/big"

func main() {
	runtime.GOMAXPROCS(10)
	var wg sync.WaitGroup
	data, err := readJson("./lol_hero.json")
	if err != nil {
		fmt.Println("失败")
	}
	func() {
		for _, name := range data {
			wg.Add(1)
			go getHeroImageIds(name, &wg)
		}
	}()
	wg.Wait()
}

func readJson(fileName string) (data map[int]string, err error) {
	body, err := ioutil.ReadFile(fileName)
	if err != nil {
		fmt.Println("读取文件出错")
		return nil, nil
	}
	err = jsoniter.Unmarshal(body, &data)
	if err != nil {
		fmt.Println("转换文件出错，", err)
		return nil, nil
	}
	return data, nil
}

func requests(url string) (body []byte) {
	client := &http.Client{}
	request, _ := http.NewRequest("GET", url, nil)
	if response, reqErr := client.Do(request); reqErr != nil {
		defer response.Body.Close()
		fmt.Printf("请求出错, 错误原因为 %s", reqErr)
		panic("请求出错")
	} else {
		defer response.Body.Close()
		body, _ := ioutil.ReadAll(response.Body)
		return body
	}
}

func getHeroImageIds(heroName string, wg *sync.WaitGroup) {
	var wg2 sync.WaitGroup
	defer func() {
		wg.Done()
		err := recover()
		if err != nil {
			fmt.Println(err)
			return
		}
	}()
	var skinList []map[string]interface{}
	heroJsUrl := fmt.Sprintf("https://lol.qq.com/biz/hero/%s.js", heroName)
	body := requests(heroJsUrl)
	com := regexp.MustCompile(`"skins":(.*?),"info"`)
	temp := com.FindSubmatch(body)
	err := jsoniter.Unmarshal(temp[1], &skinList)
	checkErr(err, "解压js文件出错")
	for _, elem := range skinList {
		skinName := elem["name"].(string)
		skinId := elem["id"].(string)
		fmt.Printf("已得到: %s-%s, 开始保存文件。。。\n", heroName, skinName)
		wg2.Add(1)
		go downloadAndSaveImages(skinId, skinName, heroName, &wg2)
	}
	wg2.Wait()
}

func checkErr(err error, msg string) {
	if err != nil {
		fmt.Println(err, msg)
		panic(msg)
	}
}

func replaceName(srcName string) (targetName string) {
	compile := regexp.MustCompile(`[\\/:*?"<>|]`)
	targetName = compile.ReplaceAllString(srcName, "")
	return
}

func downloadAndSaveImages(id, skinName, heroName string, wg2 *sync.WaitGroup) {
	defer func() {
		wg2.Done()
		err := recover()
		if err != nil {
			fmt.Println(err)
			return
		}
	}()
	heroSkinUrl := baseUrl + id + ".jpg"
	skinName = replaceName(skinName)
	imageDir := fmt.Sprintf("./lolimg/%s", heroName)
	err := os.MkdirAll(imageDir, 0644)
	checkErr(err, fmt.Sprintf("创建文件夹%s失败\n", heroName))
	file, err := os.OpenFile(imageDir+"/"+skinName+".jpg", os.O_CREATE|os.O_WRONLY, 0644)
	checkErr(err, fmt.Sprintf("创建文件%s失败\n", skinName+".jpg"))
	body := requests(heroSkinUrl)
	_, err = file.Write(body)
	checkErr(err, "图片写入文件失败\n")
	_ = file.Close()
	fmt.Println(heroName, "-", skinName, "保存成功")
}
